### PR TITLE
[mlir][ArmSME] Rewrite illegal `shape_casts` to `vector.transpose` ops

### PR DESCRIPTION
**Pull request by MacDue**
This adds a rewrite that converts illegal 2D unit-dim `shape_casts` into `vector.transpose` ops.

E.g.

```mlir
// Case 1:
%a = vector.shape_cast %0 : vector<[4]x1xf32> to vector<1x[4]xf32>
// Case 2:
%b = vector.shape_cast %1 : vector<[4]x1xf32> to vector<[4]xf32>
```

Becomes:

```mlir
// Case 1:
%a = vector.transpose %0 : [1, 0] vector<[4]x1xf32> to vector<1x[4]xf32>
// Case 2:
%t = vector.transpose %1 : [1, 0] vector<[4]x1xf32> to vector<1x[4]xf32>
%b = vector.shape_cast %t : vector<1x[4]xf32> to vector<[4]xf32>
```

Various lowerings and drop unit-dims patterns add such shape_casts, however, if they do not cancel out (which they likely won't if we've reached the vector-legalization pass) they will prevent lowering the IR.

Rewriting them as a transpose gives `LiftIllegalVectorTransposeToMemory` a chance to eliminate the illegal types.

### CODE REVIEW

**Comment by c-rhodes** [comment]
> ```mlir
> // Case 1:
> %a = vector.shape_cast %0 : vector<[4]x1xf32> to vector<1x[4]xf32>
...
> 
> Various lowerings and drop unit-dims patterns add such shape_casts, however, if they do not cancel out (which they likely won't if we've reached the vector-legalization pass) they will prevent lowering the IR.

something not clear to me here, `Case 1` would be introduced by `TransposeOpLowering` in `-convert-vector-to-llvm` and vector legalization is way before then? Or is the ordering slightly different in IREE?

**Comment by MacDue** [comment]
> > ```mlir
> > // Case 1:
> > %a = vector.shape_cast %0 : vector<[4]x1xf32> to vector<1x[4]xf32>
> > ```
> 
> ...
> 
> > Various lowerings and drop unit-dims patterns add such shape_casts, however, if they do not cancel out (which they likely won't if we've reached the vector-legalization pass) they will prevent lowering the IR.
> 
> something not clear to me here, `Case 1` would be introduced by `TransposeOpLowering` in `-convert-vector-to-llvm` and vector legalization is way before then? Or is the ordering slightly different in IREE?

In IREE the generic vector lowering happens quite a bit earlier than the ArmSME pipeline (and it's not part of `-convert-vector-to-llvm`, it's just a general vector dialect level pass).

File: mlir/lib/Dialect/ArmSME/Transforms/VectorLegalization.cpp [hunk 1]
```diff
@@ -556,6 +559,59 @@ struct LiftIllegalVectorTransposeToMemory
...
+  LogicalResult matchAndRewrite(vector::ShapeCastOp shapeCastOp,
+                                PatternRewriter &rewriter) const override {
+    auto sourceType = shapeCastOp.getSourceVectorType();
+    auto resultType = shapeCastOp.getResultVectorType();
+    if (isLegalVectorType(sourceType) || !isLegalVectorType(resultType))
+      return rewriter.notifyMatchFailure(shapeCastOp,
+                                         kMatchFailureNotIllegalToLegal);
+
+    // Note: If we know the that is source is an illegal vector type (and 2D)
+    // then dim 0 is scalable and dim 1 is fixed.
```
Comment (c-rhodes) [inline mlir/lib/Dialect/ArmSME/Transforms/VectorLegalization.cpp:595]:
```suggestion
    // Note: If we know that source is a legal vector type (and 2D)
    // then dim 0 is scalable and dim 1 is fixed.
```
?

**Comment by c-rhodes** [comment]
> > > ```mlir
> > > // Case 1:
> > > %a = vector.shape_cast %0 : vector<[4]x1xf32> to vector<1x[4]xf32>
> > > ```
> > 
> > 
> > ...
> > > Various lowerings and drop unit-dims patterns add such shape_casts, however, if they do not cancel out (which they likely won't if we've reached the vector-legalization pass) they will prevent lowering the IR.
> > 
> > 
> > something not clear to me here, `Case 1` would be introduced by `TransposeOpLowering` in `-convert-vector-to-llvm` and vector legalization is way before then? Or is the ordering slightly different in IREE?
> 
> In IREE the generic vector lowering happens quite a bit earlier than the ArmSME pipeline (and it's not part of `-convert-vector-to-llvm`, it's just a general vector dialect level pass).

ok, thanks for clarifying

**Review by c-rhodes — APPROVED** [review]
one minor comment otherwise LGTM cheers

File: mlir/lib/Dialect/ArmSME/Transforms/VectorLegalization.cpp (hunk 1 above)
Comment (MacDue) [reply mlir/lib/Dialect/ArmSME/Transforms/VectorLegalization.cpp]: Fixed the typos :+1:, but the source vector is an illegal vector type.

**Pull request merged**